# 1.0 Python Basic


In [1]:
ls

datalab@  sample_data/



## 1.1 Programming And Data Science

Data science involves processing, analyzing, and visualizing data. While some tools like Microsoft Excel allow us to perform basic data science tasks, they're limited to the functionality built in to the user interface. If you want to work with datasets that aren't structured like a spreadsheet or create entire new data visualizations from scratch, you'll need to become proficient in **programming**. Instead of using a program written by others that can solve a narrow set of tasks, you can create your own programs that can solve your specific problems.

Programming involves organizing a collection of instructions into a program for a computer to carry out. To express these instructions, we use a **programming language**. In this track, we focus on a language called Python. **Python** is a general-purpose programming language that is becoming more and more popular for doing **data science**. Companies worldwide are using Python to harvest insights from their data and get a competitive edge.


**Python** is a popular choice for working with data because it has good support for:

- handling large datasets
- working with common mathematical functions
- creating powerful data visualizations


What you will learn
- Python versions
- Expressions
- Data types
- List and Operations
- Files and Loops
- Boolean and Statements
- Challenge




### 1.1.1 Python version

There are currently two different supported versions of Python, 2.7 and 3.6. Python 3.x introduced many backwards-incompatible changes to the language, so code written for 2.7 may not work under 3.x and vice versa. For this class all code will use Python 3.6. 

https://wiki.python.org/moin/Python2orPython3

"Python 2.x is legacy, Python 3.x is the present and future of the language""



In [0]:
import sys
print(sys.version)

3.6.3 (default, Oct  3 2017, 21:45:48) 
[GCC 7.2.0]


### 1.1.2 Expressions

To introduce the basic concepts in Python, we'll start by understanding how to evaluate basic **expressions**. If you've ever used a calculator, you're familiar with the process of writing and running mathematical expressions like **4 + 5** or **(1 + 2 + 3) / 3**. Your calculator evaluates what you entered and immediately returns the result.



In [0]:
4+5

9

In [0]:
(1+2+3)/3

2.0

Python has multiple arithmetic operators that allow you to express calculations between values. In the following diagram is a list of the main arithmetic operators, some simple expressions that use each operator, and the result of each expression if we ran it in the Python console.

<center>
<img width="600" src="https://drive.google.com/uc?export=view&id=1PegHzSWBTzLRFDwoizeuJ1YqCzDGAEgn">
</center>

In [0]:
exp_1 = 5 + 5 + 5 / 3
exp_2 = (5 + 5 + 5) / 3

print("Exp #1: {0:.2f} \nExp #2: {1:.2f}".format(exp_1,exp_2))

Exp #1: 11.67 
Exp #2: 5.00


Assuming we'd explore parts of a dataset on U.S. crime rates. For each city in this dataset, the number of violent crimes that occurred in 2012 per 100,000 people are listed next to the city. Here's what the first 5 rows of that dataset look like:

| City | Crime rates  |
|-------------|------|
| Albuquerque | 749  |
| Anaheim     | 371  |
| Anchorage   | 828  |
| Arlington   | 503  |
| Atlanta     | 1379 |

Let's calculate the average of these 5 violent crime rates.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Calculates the average of **749, 371, 828, 503, 1379** and assign it to variable **crime_rates_avg**. 
2. Display the result using the print() function.



In [2]:
# put your code here
crime_rates_avg = (749+371+828+503+1379)/5
print(crime_rates_avg)


766.0



### 1.1.3 Data types

We've been working with whole numbers like **1379**, which are known as **integers**. The two most common numerical types in Python are **integer** and **float**, which is used to represent fractional (or decimal) values. **3.5** and **4.1111** are both examples of float values.

The most common non-numerical type is a **string**, which is used to represent text. To represent a piece of text as a string value, surround the text with either single quotes (**'**) or double quotes (**"**). **'Hello'** and **"Hello World!"** are both examples of **string** values.

In [0]:
print("Hello")

Unlike variable names, strings can contain special characters and spaces. You can assign a string to a variable in the same way you'd assign it a numeric value:

In [0]:
atlanta = "Atlanta"
atlanta

You may have noticed a pattern here. **Numerical values like integers and floats don't require quotation marks, but strings do**. The way in which you enter a value tells Python what data type it is. Python will use the data type to determine how the value should be handled. For example, Python allows integer variables to be divided, but not string variables. We'll learn more about that soon, but first let's practice some of the concepts you've learned so far.



**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">
1. Assign the string value **"Atlanta"** to **atlanta_string**.
2. Assign the float value **1379.5** to **atlanta_float**.
3. Print all variables using print(). 

In [3]:
# put your code here
atlanta_string = "Atlanta"
atlanta_float = 1379.5
print(atlanta_string)
print(atlanta_float)

Atlanta
1379.5


#### 1.1.3.1 The Type function

We can look up the data type of a variable's value using the **type()** function. Similar to the **print()** function, you pass a value (or variable) into the parentheses. Unlike the **print()** statement, however, the **type()** statement won't display anything. Instead, it will return the data type as a value, which can be assigned to another variable or displayed using the **print()** statement:

In [0]:
hello = 'Hello'
hello_type = type(hello)
print(hello_type)

This will return the string **class 'str'**, which means that the value associated with **hello** is a string (**str** is short for string). To avoid having to create a variable each time, you can chain the **print()** and **type()** statements:

In [0]:
hello = 'Hello'
print(type(hello))

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">
1. Display the type of **atlanta_string**.

In [4]:
# put your code  here
print(type(atlanta_string))

<class 'str'>


## 1.2 Using A List To Store Multiple Values

So far, we've been storing individual values in variables. Often in data science, we're working with thousands of data points that are grouped together in a certain way and have an order to them. We need a container that can hold multiple values that we can use to perform operations on.

We can use a **list**, which is an object that represents a sequence of values. For example, we can represent the cities in our dataset as a list as a sequence of strings (**"Albuquerque", "Anaheim"**, and so on).

<img width="300" src="https://drive.google.com/uc?export=view&id=1zsaj8TYl5PgLmkTFrWhhXprhmDY5HPqm">


The most basic way to make a list is to create an empty one first, and then adding values to it. To create an empty list, assign a pair of empty brackets **[ ]** to a variable:


In [0]:
cities = []

To add values to a list object, use the **list.append()** method. This method accepts a value and adds it to a list object. Unlike functions, methods are called using dot notation **( . )** on a specific object.

Because **cities** is a list object, the Python interpreter knows that **list.append()** can be used. You can see the methods available to list objects [here](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists).

In the following code snippet, we use the **list.append()** method to add the string **"Albuquerque"** then the string **"Anaheim"** to the list cities:

In [0]:
cities.append("Albuquerque")
cities.append("Anaheim")
print(cities)
print(type(cities))

Each time we call **list.append()**, the values in the **list cities** are updated.

<img width="500" src="https://drive.google.com/uc?export=view&id=16c7RFrqN9HKdXClIfEAQHn1GBYEcB02-">


Lastly, **list objects can store values of multiple types**. In the following code example, we append a city name first then its violent crime rate:

In [0]:
cities = []
cities.append("Albuquerque")
cities.append(749)
cities.append("Anaheim")
cities.append(371)

print(cities)

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create two empty lists, **cities** and **crime_rates**.
2. Add the following strings to the list **cities**, in the following order:
   - "Albuquerque"
   - "Anaheim"
   - "Anchorage"
   - "Arlington"
   - "Atlanta"
3. Add the following integer values to the list **crime_rates**, in the following order:
  - 749
  - 371
  - 828
  - 503
  - 1379
4. Use the **print()** function to display both lists.

In [5]:
# put your code here
cities = []
crime_rates = []

cities.append("Albuquerque")
cities.append("Anaheim")
cities.append("Anchorage")
cities.append("Arlington")
cities.append("Atlanta")

crime_rates.append(749)
crime_rates.append(371)
crime_rates.append(828)
crime_rates.append(503)
crime_rates.append(1379)

print(cities)
print(crime_rates)


['Albuquerque', 'Anaheim', 'Anchorage', 'Arlington', 'Atlanta']
[749, 371, 828, 503, 1379]


### 1.2.1 Creating Lists With Values

In the last step, we learned how to populate lists by:

- creating an empty list: **months = [ ]**
- then adding values to the list using the **list.append()** method: **months.append(1)**


This can become tedious when working with multiple lists, because you have to write many lines of code (one for each value in the list). You can instead create a list and populate a values all in one line using the following syntax:

>```python
cities = ["Albuquerque", "Anaheim", "Anchorage", "Arlington", "Atlanta"]
```

You may be wondering why we would ever want to create an empty list and add values manually. That technique is useful when you only want to add values that meet specific criteria. In that case, you need an empty placeholder list that you can append items to individually. As you work with lists, you'll start to learn the best technique for a given situation.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create the list **crime_rates** again, but this time using the technique we just learned.
2. This list should contain the following values in the following order:
  - 749
  - 371
  - 828
  - 503
  - 1379

In [7]:
# put your code here
crime_rates = [749, 371, 828, 503, 1379]
print(crime_rates)

[749, 371, 828, 503, 1379]


### 1.2.2 Accessing Elements In A List


Now that we know how to create a list and add values to it, let's learn how to access and work with the values in a list we've created. Each value in the list has an **index**, or position, associated with it. A list starts at index **0** and goes all the way to one less than the total number of values, or elements, in the list. If you have a list with **5** values, for example, the indexes will range from **0 to 4**.

The main quirk of list indexes is that to access the first element in a list, we actually use the index value 0 - not 1. The second element is accessed with index value 1, the third element with index value 2, and so on. This is known as **zero indexing**. While many programming languages use zero indexing, some, like MATLAB, do not.

<img width="300" src="https://drive.google.com/uc?export=view&id=16lnTG7VA6CmsqQkrMhBHOUSV5TWEynJO">

To return the value that has a given index, pass the integer for the index into bracket notation. In the following code, we access the first, second, and fifth elements in the list **crime_rates**. We assign each of the accessed values to new variables:

>```python
crime_rates = [749, 371, 828, 503, 1379]
first_value = crime_rates[0]
second_value = crime_rates[1]
fifth_value = crime_rates[4]
```

The Python interpreter expects that the bracketed integer value will be within the list's range of indexes. Passing in a non-integer value or an integer value outside of the range of indexes (e.g. index 7 for a list only containing 5 elements) will result in an error.


In [0]:
crime_rates = [749, 371, 828, 503, 1379]
crime_rates[7]

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Select the third element from the list **cities** and assign to the variable **anchorage_str**.
2. Select the third element from the list **crime_rates** and assign to the new variable **anchorage_cr**.


In [8]:
cities = ["Albuquerque", "Anaheim", "Anchorage", "Arlington", "Atlanta"]
crime_rates = [749, 371, 828, 503, 1379]

# put your code here
anchorage_str = cities[2]
anchorage_cr = crime_rates[2]

print(anchorage_str, anchorage_cr)

Anchorage 828



### 1.2.3 Retrieving The Length Of A List

We mentioned earlier that trying to lookup a value at an index that's not in the list will return an error and cause your code to halt. You may be wondering how we avoid accidentally looking up a value that's outside the index of a list. Python's **len()** function returns the **length** of a list, or the number of elements in that list. The function returns this value as an integer:

>```python
crime_rates = [749, 371, 828, 503, 1379]
len_cr = len(crime_rates)  # Points to the integer value 5.
```

If we're ever unsure about the number of elements in a list, we can pass the list into the **len()** function. Because the **len()** function returns an integer, we can subtract 1 from this number to retrieve the index of the last element.



>```python
crime_rates = [749, 371, 828, 503, 1379]
last_index = len(crime_rates) - 1 # last_value contains the value at index 4 (the last value in the list).
last_value = crime_rates[last_index]
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Add the lengths of the **cities** and **crime_rates** list objects and assign the sum to **two_sum**.


In [9]:
cities = ["Albuquerque", "Anaheim", "Anchorage", "Arlington", "Atlanta"]
crime_rates = [749, 371, 828, 503, 1379]

# put your code here
two_sum = len(cities) + len(crime_rates)

print(two_sum)


10


### 1.2.4 Slicing lists

If we have a list containing thousands of values and want to retrieve the ones between index **10** and **500**, this would be a lot of work with what we know so far. Fortunately, lists have a feature called **slicing** that allows you to return all of the values between a starting index and an ending index. When you slice a list, you return a new list containing just the values you're interested in. The value at the starting index and all of the values in between will be returned. The value at the ending index will not.

To slice a list, pass the starting and ending index positions into the brackets as integer values, separated by a colon ** : **. In the following code, we use the slice 2:4 to return a new list containing the values at indices 2 and 3:

In [0]:
crime_rates = [749, 371, 828, 503, 1379]
# The following slice selects values at index 2 and 3, but not 4.
two_four = crime_rates[2:4]
two_four

Here's a diagram of the same slice:

<img width="300" src="https://drive.google.com/uc?export=view&id=1GB0hKIY15h0mjM5jlJKKRXJhlCIaPct9">


In the following code, we use the **len()** function to retrieve the total number of elements in the list **crime_rates**, and use it as the ending index:

In [0]:
# Values at index 2, 3, 4.
ending_index = len(crime_rates)
two_to_end = crime_rates[2:ending_index] #or crime_rates[2:]
two_to_end

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Select the second, third, and fourth elements from **cities**, and assign the resulting list to **cities_slice**.
2. Select the last two elements in **crime_rates** and assign to **cr_slice**.

In [11]:
cities = ["Albuquerque", "Anaheim", "Anchorage", "Arlington", "Atlanta"]
crime_rates = [749, 371, 828, 503, 1379]

# put your code here
cities_slice = cities[1:4]
cr_slice = crime_rates[3:]
print(cities_slice, cr_slice)

['Anaheim', 'Anchorage', 'Arlington'] [503, 1379]


# 2.0 Files and Loops



In this section, we'll learn how to work with files and use loops to iterate through lists. We'll be working with crime rate data for 73 cities in the United States. 

**Datasets** are often represented in files that you can download and manipulate. Before we get started, we'll first need to learn how to work with files in Python.

To open a file in Python, we use the **open()** function. This function accepts two different arguments (inputs) in the parentheses, always in the following order:

- the name of the file (as a string)
- the mode of working with the file (as a string)


We'll learn about the various modes in a later mission. For now, we'll just use **"r"**, the mode for reading in files.

When entering multiple inputs, separate them with commas (** , ** ). For example, to open a file named **story.txt** in read mode, we write the following:

>```python
open("story.txt", "r")
```

The **open()** function returns a *File* object. This object stores the information we passed in, and allows us to call methods specific to the File class. We can assign the File object to a variable so we can refer to it later:

>```python
a = open("story.txt", "r")
```

Note that the File object, **a**, won't contain the actual contents of the file. It's instead an object that acts as an interface to the file and contains methods for reading in and modifying the file's contents (which we'll cover in the next screen).




### 2.1 Uploading files from your local file system

In [6]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving crime_rates.csv to crime_rates.csv
Saving dq_unisex_names.csv to dq_unisex_names.csv
Saving la_weather.csv to la_weather.csv
User uploaded file "crime_rates.csv" with length 1030 bytes
User uploaded file "dq_unisex_names.csv" with length 10497 bytes
User uploaded file "la_weather.csv" with length 3305 bytes


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Use the **open()** function to create a File object.
  - The name of the file to open is **"crime_rates.csv"**. 
  - Access the file in read mode (** "r" **).
2. Assign this File object to the variable **f**.

In [0]:
# put your code here
f = open("crime_rates.csv", "r")

### 2.2 Reading in Files

File objects have a **read()** method that returns a string representation of the text in a file. Unlike the **append()** method from the previous mission, the **read()** method returns a value instead of modifying the object that calls the method. In the following code, we use the **read()** function to read the contents of **"test.txt"** into a File object, and assign that object to **g**:

>```python
f = open("test.txt", "r")
g = f.read()
```

Since **g** is a string, we can use the **print()** function to display the contents of the file:

>```python
f = open("test.txt", "r")
g = f.read()
print(g)
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Run the **read()** method on the File object **f** to return the string representation of **crime_rates.csv**.
2. Assign the resulting string to a new variable named **data**.


In [11]:
# put your code here
data = f.read()
print(data)

Albuquerque,749
Anaheim,371
Anchorage,828
Arlington,503
Atlanta,1379
Aurora,425
Austin,408
Bakersfield,542
Baltimore,1405
Boston,835
Buffalo,1288
Charlotte-Mecklenburg,647
Cincinnati,974
Cleveland,1383
Colorado Springs,455
Corpus Christi,658
Dallas,675
Denver,615
Detroit,2122
El Paso,423
Fort Wayne,362
Fort Worth,587
Fresno,543
Greensboro,563
Henderson,168
Houston,992
Indianapolis,1185
Jacksonville,617
Jersey City,734
Kansas City,1263
Las Vegas,784
Lexington,352
Lincoln,397
Long Beach,575
Los Angeles,481
Louisville Metro,598
Memphis,1750
Mesa,399
Miami,1172
Milwaukee,1294
Minneapolis,992
Mobile,522
Nashville,1216
New Orleans,815
New York,639
Newark,1154
Oakland,1993
Oklahoma City,919
Omaha,594
Philadelphia,1160
Phoenix,636
Pittsburgh,752
Plano,130
Portland,517
Raleigh,423
Riverside,443
Sacramento,738
San Antonio,503
San Diego,413
San Francisco,704
San Jose,363
Santa Ana,401
Seattle,597
St. Louis,1776
St. Paul,722
Stockton,1548
Tampa,616
Toledo,1171
Tucson,724
Tulsa,990
Virginia Beach,1

### 2.3 Splitting

To make our string object data more useful, let's convert it into a **list**. Here's a preview of how the dataset looks:

```python
f = open("test.txt", "r")
g = f.read()
```

```python
Albuquerque,749\nAnaheim,371\nAnchorage,828\n
```

Each line is separated by the string **\n**, which is referred to as the **new-line character**. When we open a text file in a text editor, the editor will automatically split the text and create a new line wherever it sees the string **\n**:

<center>
<img width="200" src="https://drive.google.com/uc?export=view&id=1tOua2vNM4dFMS5VdMs1CpoZtX0fMcugu">
</center>
  
In Python, we can use the **split()** method to turn a string object into a list of strings, like so:

```python
f = open("test.txt", "r")
g = f.read()
print(g)
```

```python
["Albuquerque,749", "Anaheim,371", "Anchorage,828"]
```

The **split()** method takes a string input corresponding to the delimiter, or separator. This delimiter determines how the string is split into elements in a list. For example, the delimiter for the crime rate data we just looked at is **\n**. Many other files use commas to separate elements:

>```python
sample = "john,plastic,joe"
split_list = sample.split(",") # split_list is a list of _strings_: ["john", "plastic", "joe"]
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Split the string object **data** on the new-line character **"\n"**
2. Store the result in a variable named **rows**. 
3. Then, use the **print()** function to display the first five elements in rows.

In [21]:
# We can split a string into a list.
sample = "john,plastic,joe"
split_list = sample.split(",")
print(split_list)

# Here's another example.
string_two = "How much wood\ncan a woodchuck chuck\nif a woodchuck\ncould chuck wood?"
split_string_two = string_two.split('\n')
print(split_string_two)

# Code from previous cells
f = open('crime_rates.csv', 'r')
data = f.read()

# Put your code here
rows = data.split("\n")
print(rows)

['john', 'plastic', 'joe']
['How much wood', 'can a woodchuck chuck', 'if a woodchuck', 'could chuck wood?']
['Albuquerque,749', 'Anaheim,371', 'Anchorage,828', 'Arlington,503', 'Atlanta,1379', 'Aurora,425', 'Austin,408', 'Bakersfield,542', 'Baltimore,1405', 'Boston,835', 'Buffalo,1288', 'Charlotte-Mecklenburg,647', 'Cincinnati,974', 'Cleveland,1383', 'Colorado Springs,455', 'Corpus Christi,658', 'Dallas,675', 'Denver,615', 'Detroit,2122', 'El Paso,423', 'Fort Wayne,362', 'Fort Worth,587', 'Fresno,543', 'Greensboro,563', 'Henderson,168', 'Houston,992', 'Indianapolis,1185', 'Jacksonville,617', 'Jersey City,734', 'Kansas City,1263', 'Las Vegas,784', 'Lexington,352', 'Lincoln,397', 'Long Beach,575', 'Los Angeles,481', 'Louisville Metro,598', 'Memphis,1750', 'Mesa,399', 'Miami,1172', 'Milwaukee,1294', 'Minneapolis,992', 'Mobile,522', 'Nashville,1216', 'New Orleans,815', 'New York,639', 'Newark,1154', 'Oakland,1993', 'Oklahoma City,919', 'Omaha,594', 'Philadelphia,1160', 'Phoenix,636', 'Pit

### 2.4 Loops

We now have a list representation of the dataset (**rows**). Each element in the dataset is a string containing a comma (** , ** ) that separates the city name from the crime rate. Because they're strings, we can use the **split()** method on each of them to separate those values. Accomplishing this based on what we know so far would be a cumbersome task. **We would need to write 73 lines of code, one for each element in the list**.

Ideally, we could specify the code we want to execute for each element in the list a single time. Python and most other programming languages allow you to do this with a **loop**. For example, here's a loop that prints each element in a list:

>```python
cities = ["Austin", "Dallas", "Houston"]
for city in cities:
    print(city)
```


Here are the steps Python takes when it runs this code:

- Assigns the value at index **0** in the **cities** list (**"Austin"**) to **city**.
- Executes the indented code. Because the current value of **city** is **"Austin"**, it runs **print("Austin")**.
- Assigns the value at index **1** in the **cities** list (**"Dallas"**) to **city**.
- Executes the indented code. Because the current value of **city** is **"Dallas"**, it runs **print("Dallas")**.
- Assigns the value at index **2** from the **cities** list (**"Houston"**) to **city**.
- Executes the indented code. Because the current value of **city** is **"Houston"**, it runs **print("Houston")**.
- Sees that there are no more elements in the list **cities**, and stops running.

This code uses a type of loop called a **for loop**. We can break the for loop down into two main components: 
- the for loop itself, 
- and the loop body that contains the code we want to run during each iteration.

Here are the things you need to include in each of these components:

- for Loop

  - **Syntax** - the words **for** and **in** need to be included in the statement
  - **Iterator variable** - the variable name you decide to use to refer to each element in the list (**city**)
  - **Sequence** - the variable you want to iterate over (**cities**)
  - **Colon** - loop statements must end with a colon (**:**)
  
- Loop Body

  - **Indentation** - every line of code within the loop should be indented four spaces
  - **Logic** - the actual code we want to execute for each element. In the above code block, for example, we update the iterator variable (**city**) in each iteration of the loop
  
  
  Here's a diagram displaying the values for **city** and the **print()** statement during each iteration of the loop:



<img width="300" src="https://drive.google.com/uc?export=view&id=1wafTcaMqGbnkCw80aKKz73CCNyQym22L">

The iterator variable, **city**, is a temporary variable that's only accessible within the for loop.

The very basic loop body we wrote above only contains one line of code. To write a loop body with multiple lines of code, we need to indent that code consistently (using four spaces). For example, the following code will run the **print()** statement three times for each element in **cities**:

>```python
for city in cities:
    print(city)
    print(city)
    print(city)
```

Here's an annotated diagram of the code:

<img width="400" src="https://drive.google.com/uc?export=view&id=1fbwcQWaQAgmu-7TYKEWBQSYJ6HjDC217">


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. The variable **ten_rows** contains the first 10 elements in rows.
2. Write a for loop that:
  - iterates over each element in **ten_rows**
  - uses the **print()** function to display each element

In [26]:
ten_rows = rows[0:10]

# put your code here
for rows in ten_rows:
  print(rows)

Albuquerque,749
Anaheim,371
Anchorage,828
Arlington,503
Atlanta,1379
Aurora,425
Austin,408
Bakersfield,542
Baltimore,1405
Boston,835


### 2.5 Lists of Lists

Now that we know how to execute code for each element in a list, we can use a loop to split those elements on a delimiter and append the results to a new list. So far, we've appended values like integers and strings, but we've never appended a list to a list. This is known as a **list of lists**, since each element is itself a list.

The following code:

- splits each element in **three_rows** (which contains the first three elements from rows) on the comma delimiter (** , **)
- appends the resulting list (**split_list**) to a new list we create (**final_list**)
- displays the final list using the **print()** function

>```python
three_rows = ["Albuquerque,749", "Anaheim,371", "Anchorage,828"]
final_list = []
for row in three_rows:
    split_list = row.split(',')
    final_list.append(split_list)
print(final_list)
```
Here's the logic above represented as a diagram:

<img width="600" src="https://drive.google.com/uc?export=view&id=1YoONZjbomKK8DgnAxn22nuDEMhIqL0e7">


In the leftmost part of the diagram, you'll notice that the elements in **three_rows** are strings. In the right-most part of the diagram, you'll notice that the elements in **final_list** are lists.

Here's what the output looks like:

>```python
[['Albuquerque', '749'], ['Anaheim', '371'], ['Anchorage', '828']]
```

Recall that since only the indented code is executed as part of the loop, the unindented **print()** statement on the last line will execute only after the loop finishes. In the following code cell, we execute the code we visualized in the diagram, along with some additional **print()** statements. These **print()** statements retrieve individual elements from **final_list**. Note that they all display as list objects. You'll practice writing your own for loop in the next step. For now, explore and run the code we covered in this step in the code cell below.


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Write a **for loop** that splits each element in rows on the comma delimiter, and appends the resulting list to a new list named **final_data**.
2. Then, use the **print()** function and list slicing to display the first five elements in **final_data**.

In [29]:
f = open('crime_rates.csv', 'r')
data = f.read()
rows = data.split('\n')
print(rows[0:5])

#put your code here
final_data = []
for x in rows:
  final_data.append(x.split(","))
print(final_data)

['Albuquerque,749', 'Anaheim,371', 'Anchorage,828', 'Arlington,503', 'Atlanta,1379']
[['Albuquerque', '749'], ['Anaheim', '371'], ['Anchorage', '828'], ['Arlington', '503'], ['Atlanta', '1379'], ['Aurora', '425'], ['Austin', '408'], ['Bakersfield', '542'], ['Baltimore', '1405'], ['Boston', '835'], ['Buffalo', '1288'], ['Charlotte-Mecklenburg', '647'], ['Cincinnati', '974'], ['Cleveland', '1383'], ['Colorado Springs', '455'], ['Corpus Christi', '658'], ['Dallas', '675'], ['Denver', '615'], ['Detroit', '2122'], ['El Paso', '423'], ['Fort Wayne', '362'], ['Fort Worth', '587'], ['Fresno', '543'], ['Greensboro', '563'], ['Henderson', '168'], ['Houston', '992'], ['Indianapolis', '1185'], ['Jacksonville', '617'], ['Jersey City', '734'], ['Kansas City', '1263'], ['Las Vegas', '784'], ['Lexington', '352'], ['Lincoln', '397'], ['Long Beach', '575'], ['Los Angeles', '481'], ['Louisville Metro', '598'], ['Memphis', '1750'], ['Mesa', '399'], ['Miami', '1172'], ['Milwaukee', '1294'], ['Minneapolis',

### 2.6 Accessing Elements in a List of Lists: The Manual Way

A list of lists has some unique interaction mechanisms. Using bracket notation to retrieve an element at a certain index returns a list object. However, using bracket notation on the resulting list will actually return a data point (such as a string or an integer).

In the following code, we retrieve the first list from **final_data** and assign it to **first_list**. We then retrieve the first element from **first_list** and assign it to **first_list_first_value**:

>```python
first_list = final_data[0] # Returns the first list: ['Albuquerque', '749'].
first_list_first_value = first_list[0] # Returns the first list's first element: 'Albuquerque'.
```

Since using bracket notation once returns a list, you can use bracket notation again on the resulting list without having to assign it to a variable first. Using double bracket notation is simpler. The following code uses double bracket notation to retrieve the value at index **0** for the first three elements in **final_data**:

>```python
first_list_first_value = final_data[0][0] # Returns the first list's first element, 'Albuquerque'.
second_list_first_value = final_data[1][0] # Returns the second list's first element, 'Anaheim'.
third_list_first_value = final_data[2][0] # Returns the third list's first element, 'Anchorage'.
```

Here's a visual representation of **final_data[0][0]** to help you grasp double bracket notation:

<img width="600" src="https://drive.google.com/uc?export=view&id=1YxyDUoGE1KrjsKB65S8iJ5CFhUiFCDnS">



**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create the variable **five_elements** that contains the first five elements from **final_data**.
2. Create a list of strings named **cities_list** that contains the city names from each list in **five_elements**.

In [36]:
# put your code here
five_elements = final_data[:5]
cities_list = []
cities_list.append(five_elements[0][0])
cities_list.append(five_elements[1][0])
cities_list.append(five_elements[2][0])
cities_list.append(five_elements[3][0])
cities_list.append(five_elements[4][0])
print(cities_list)

['Albuquerque', 'Anaheim', 'Anchorage', 'Arlington', 'Atlanta']


### 2.7 Looping Through a List of Lists

Instead of appending each element from the nested list to a new list individually, we can use a for loop and specify the append logic within the loop body. Let's explore how we can replace the **append()** statements from the previous step with a for loop.

In the following code, we create a new list called **crime_rates**, and write a for loop that iterates through **five_elements**. In each iteration of the loop, we retrieve the string at index **1** from the current list (within **five_elements**), assign that string to the variable **crime_rate**, and use the **append()** method to add **crime_rate** to the list **crime_rates**.

>```python
crime_rates = []
for row in five_elements:
    crime_rate = row[1] # row is a list variable, not a string.
    crime_rates.append(crime_rate)  # crime_rate is a string, the crime rate of the city
```

Now it's your turn to apply what you've learned to the full dataset.


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create a list of strings named **cities_list** that contains just the city names from **final_data**.
  - Recall that the city name is located at index **0** for each list in **final_data**.

In [38]:
cities_list = []
for x in final_data:
  cities_list.append(x[0])
print(cities_list)

['Albuquerque', 'Anaheim', 'Anchorage', 'Arlington', 'Atlanta', 'Aurora', 'Austin', 'Bakersfield', 'Baltimore', 'Boston', 'Buffalo', 'Charlotte-Mecklenburg', 'Cincinnati', 'Cleveland', 'Colorado Springs', 'Corpus Christi', 'Dallas', 'Denver', 'Detroit', 'El Paso', 'Fort Wayne', 'Fort Worth', 'Fresno', 'Greensboro', 'Henderson', 'Houston', 'Indianapolis', 'Jacksonville', 'Jersey City', 'Kansas City', 'Las Vegas', 'Lexington', 'Lincoln', 'Long Beach', 'Los Angeles', 'Louisville Metro', 'Memphis', 'Mesa', 'Miami', 'Milwaukee', 'Minneapolis', 'Mobile', 'Nashville', 'New Orleans', 'New York', 'Newark', 'Oakland', 'Oklahoma City', 'Omaha', 'Philadelphia', 'Phoenix', 'Pittsburgh', 'Plano', 'Portland', 'Raleigh', 'Riverside', 'Sacramento', 'San Antonio', 'San Diego', 'San Francisco', 'San Jose', 'Santa Ana', 'Seattle', 'St. Louis', 'St. Paul', 'Stockton', 'Tampa', 'Toledo', 'Tucson', 'Tulsa', 'Virginia Beach', 'Washington', 'Wichita']


### 2.8 Challenge

In this section, we learned how to display the type associated with any value. We can actually convert between types using specific functions in Python:

- **int()**: convert to integer value
- **float()**: convert to float value
- **str()**: convert to string value

Let's focus on just the **int()** function for now. We can use this function to convert string values that represent only an integer value:

>```python
int_five = int("5")
```

If we try to convert a string value that isn't 100% an integer value just represented as a string, we'll get an error:



In [1]:
int("5.0")

ValueError: ignored



We'll end this section by providing an opportunity to synthesize the concepts you've learned so far. Note that this exercise may be more challenging than the previous ones, and may take more time to complete.


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create a list of integers named **int_crime_rates** that contains just the crime rates - as integers - from the list **rows**.

  - First create an empty list and assign it to a new variable **int_crime_rates**.
  - Then, write a for loop that iterates over rows and does the following:
    - uses the **split()** method to convert each string in **rows** into a list, based on the comma delimiter
    - converts the value at index **1** from that list to an integer using the **int()** function
    - uses the **append()** method to add each integer to **int_crime_rates**
    

In [13]:
f = open('crime_rates.csv', 'r')
data = f.read()
rows = data.split('\n')
print(rows[0:5])

#put you code here
int_crime_rates = [] #Lista so com numeros
new_rows = [] #Lista para separar nome do numero

for x in rows[:5]:
  new_rows.append(x.split(","))
  
for x in new_rows:
  int_crime_rates.append(int(x[1]))
  
print(int_crime_rates)

['Albuquerque,749', 'Anaheim,371', 'Anchorage,828', 'Arlington,503', 'Atlanta,1379']
[749, 371, 828, 503, 1379]


# 3.0 Booleans and if Statements



## 3.1. Booleans

We'll learn how to express conditional logic. We can use **conditional logic** to add criteria to the code we write. Some examples of operations that use criteria include:

- Finding all the integers in a list that are greater than **"5"**.
- Identifying which elements in a list are strings, and printing only those values.

We can break down both of these examples into logic we can code:

- For each integer in a list, if the integer is greater than **"5"**, add to the list **greater_than_five**.
- For each element in a list, if the value has a string data type, use the **print()** function to display it; if it's not a string, ignore it.

Python has a class called **Boolean** that helps express conditional logic. There are only two Boolean values: **True** and **False**. Because they're words, Boolean values may look like strings, but they're an entirely separate class. For example, string operations like concatenation won't work with Booleans.

The following code example assigns **True** to **t** and **False** to **f**:

>```python
t = True
f = False
```

If we display the data type for either **t** or **f**, we'll see class **'bool'**, shorthand for Boolean.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Assign the value **True** to the variable **cat**, and the value **False** to the variable **dog**.
2. Then, use the **print()** function and the **type()** function to display the type for **cat**.


In [15]:
# put your code here
cat = True
dog = False
print(type(True))

<class 'bool'>




## 3.2 Boolean Operators

Python has comparison operators that allow us to compare variables:

- **==** returns **True** if both variables are equivalent, and **False** if they're different
- **!=** returns **True** if both variables are different, and **False** if they're equivalent
- **>** returns **True** if the first variable is greater than the second variable, and **False** otherwise
- **<** returns **True** if the first variable is less than the second variable, and **False** otherwise
- **>=** returns **True** if the first variable is greater than or equal to the second variable, and **False** otherwise
- **<=** returns **True** if the first variable is less than or equal to the second variable, and **False** otherwise

For now let's focus on the equality operators (**==** and **!=**). To compare two variables, place the operator between them. We recommend adding a space before and after the operator for better readability:

>```python
print(8 == 8) # True
print(8 != 8) # False
print(8 == 10) # False
print(8 != 10) # True
```

We can assign the result of a comparison to a Boolean variable:

```python
# True 
t = (8 == 8) 
# False
u = (8 != 8) 
```

We can also compare strings, floats, Booleans, and even lists:

>```python
"8" == "8"
["January", "February"] == ["January", "February"]
5.0 == 5.0
```


The variable **cities** is a list of strings containing the city names from the crime rate dataset we used in the previous section. We've created the list for you already.

In [0]:
cities = ['Albuquerque','Anaheim','Anchorage','Arlington','Atlanta','Aurora',
'Austin','Bakersfield','Baltimore','Boston','Buffalo','Charlotte-Mecklenburg',
'Cincinnati','Cleveland','Colorado Springs','Corpus Christi','Dallas','Denver',
'Detroit','El Paso','Fort Wayne','Fort Worth','Fresno','Greensboro','Henderson',
'Houston','Indianapolis','Jacksonville','Jersey City','Kansas City','Las Vegas',
'Lexington','Lincoln','Long Beach','Los Angeles','Louisville Metro','Memphis',
'Mesa','Miami','Milwaukee','Minneapolis','Mobile','Nashville','New Orleans',
'New York','Newark','Oakland','Oklahoma City','Omaha','Philadelphia','Phoenix',
'Pittsburgh','Plano','Portland','Raleigh','Riverside','Sacramento','San Antonio',
'San Diego','San Francisco''San Jose','Santa Ana','Seattle','St. Louis','St. Paul',
'Stockton','Tampa','Toledo','Tucson','Tulsa','Virginia Beach','Washington','Wichita']

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Use the Boolean operators to determine if the following pairs of values are equivalent:
  - The first element in **cities** and the string **"Albuquerque"**. Assign the resulting Boolean value to **first_alb**.
  - The second element in **cities** and the string **"Albuquerque"**. Assign the resulting Boolean value to **second_alb**.
  - The first element in **cities** and the last element in **cities**. Assign the resulting Boolean value to **first_last**.

In [26]:
# put your code here

first_alb = (cities[0] == "Albuquerque")
second_alb = (cities[1] == "Albuquerque")
first_last = (cities[0] == cities[-1])

print(first_alb, second_alb, first_last)

True False False



## 3.3 Booleans with "Greater Than"

We can use the greater than operator (** > ** ) to test whether one value is larger than another. Similarly, the greater than or equal to operator (** >= ** ) determines if one value is larger than or equal to a second value:

>```python
rates = [10, 15, 20]
rates[0] > rates[1] # False
rates[0] >= rates[0] # True
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. The variable **crime_rates** is a list of integers containing the crime rates from the dataset. Perform the following comparisons:
  - Evaluate whether the first element in **crime_rates** is larger than the integer 500, and assign the Boolean result to **first_500**.
  - Evaluate whether the first element in **crime_rates** is larger than or equal to 749, and assign the Boolean result to **first_749**.
  - Evaluate whether the first element in **crime_rates** is greater than or equal to the last element in **crime_rates**, and assign the Boolean result to **first_last**.

In [2]:
crime_rates = [749,371,828,503,1379,425,408,542,1405,835,1288,647,974,1383,455,658,
675,615,2122,423,362,587,543,563,168,992,1185,617,734,1263,784,352,397,575,481,598,
1750,399,1172,1294,992,522,1216,815,639,1154,1993,919,594,1160,636,752,130,517,423,
443,738,503,413,704,363,401,597,1776,722,1548,616,1171,724,990,169,1177,742]

# put your code here

first_500 = (crime_rates[0] > 500)
first_749 = (crime_rates[0] > 749)
first_last = (crime_rates[0] >= crime_rates[-1])

print(first_500, first_749, first_last)

True False True


## 3.4 Booleans with "Less Than"

We can use the less than operator (** < **) to test whether one value is smaller than another value. Similarly, the less than or equal to operator (** <= **) determines if one value is smaller than or equal to another value:

>```python
rates = [10, 15, 20]
rates[0] < rates[1] # True
rates[0] <= rates[0] # True
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. The variable **crime_rates** is a list containing the crime rates from the dataset as integers. Perform the following comparisons:
  - Determine whether the second element in **crime_rates** is smaller than the integer 500, and assign the Boolean result to **second_500**.
  - Determine whether the second element in **crime_rates** is smaller than or equal to 371, and assign the Boolean result to **second_371**.
  - Determine whether the second element in **crime_rates** is smaller than or equal to the last element in **crime_rates**, and assign the Boolean result to **second_last**.

In [12]:
# put your code here
second_500 = crime_rates[1] < 500
second_371 = crime_rates[1] < 371
second_last = crime_rates[1] <= crime_rates[-1]

print("%r \n%r \n%r\n" % (second_500, second_371, second_last))

True 
False 
True



## 3.5 If Statements


Now that we know how to work with **Boolean** values, let's dive more into how we use Booleans to express conditional logic. To complement Booleans, Python contains the **if operator**. We can use this operator to write a statement that tests whether certain conditions exist. Our if statement will evaluate to either **True** or **False**, and only run the specified code when **True**.

For example, the following code checks whether the integer value assigned to **sample_rate** is larger than 5. This is referred to as a **"conditional statement."** It assigns the Boolean result to greater, and uses the **print()** function to display **sample_rate** if greater is **True**:

>```python
sample_rate = 749
greater = (sample_rate > 5)
if greater:                    #This is the conditional statement.
    print(sample_rate)
```


We can also specify the conditional statement inside the **if statement**:

>```python
if sample_rate > 5:            #This is the conditional statement.
    print(sample_rate)
```


The conditional statement after the if has to evaluate to either **True** or **False**.

Similar to for loops, we need to format if statements in the following way:

- End the conditional statement with a colon (** : **)
- Indent the code (that we want run when **True**) below the conditional statement

Also similar to **for loops**, if statements can contain multiple lines in the body, as long as their indentation aligns.

>```python
t = True
f = False
if t:
    print("Now you see me")
if f:
    print("Now you don't")
```

We'll end with a generalized representation of an if statement:

<img width="400" src="https://drive.google.com/uc?export=view&id=1U5rhJWdAfhC5Kzoy_MR5SFw333DUWebw">


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Determine whether the third element in **cities** is equivalent to the string **"Anchorage"**.
  - Create the variable **result** with value 0.
  - If it is, change the variable **result** to **1**.

In [29]:
# put your code here

if (cities[2] == "Anchorage"):
  result = 0
else:
  result = 1
  
print(result)

0


## 3.6 Nesting If Statements

We can **nest if statements** to specify multiple criteria. In the following code example, we first test whether an integer is greater than **500**, and then check whether it's greater than **1000.** Python only evaluates the inner if statement if the outer if statement evaluates to **True**.

>```python
value = 1500
if value > 500:
    if value > 1000:
        print("This number is HUGE!")
```

Here's an annotated diagram of the nested if statement:

<img width="400" src="https://drive.google.com/uc?export=view&id=1JPabfwhjCRwwlsDh3IXv7iw0D3nwC1RF">

Be sure to indent your code properly with four spaces for each code block!

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Write a piece of code that nests the following concepts in the order in which they appear:
  - An **if statement** that tests whether the first element in **crime_rates** is larger than 500
  - A second if statement that tests whether the second element in **crime_rates** is larger than 300
  - If both statements evaluate to **True**, assign the value **True** to the variable **both_conditions**

In [31]:
# put your code here

if crime_rates[0] > 500:
  if crime_rates[1] > 300:
     both_conditions = True
      
print( both_conditions)

True


## 3.7 If Statements and For Loops

We can also **nest if statements** within **for loops**, and vice versa. For example, we can search a list for the existence of a specific value by combining a for loop with an if statement. The if statement determines whether the current element is equivalent to the value we're interested in:

>```python
found = False
for city in cities:
    if city == 'Washington':
        found = True
```

We've set the value found set to **False** by default. If one of the elements in cities is equivalent to the string **"Washington"**, then Python assigns the Boolean value **True** to found.

We can also use a for loop and an if statement to determine the index for a specific value in a list. To accomplish this, we can create an integer variable named counter outside the for loop and increment it by 1 with each iteration. If the value at the current iteration of the loop matches our desired value, we can set another integer variable, index, to the current value of counter:


>```python
counter = 0
index = 0
for city in cities:
    if city == "Washington":
        index = counter
    counter += 1
```

Keep in mind that in the above implementation, if there are multiple instances of the desired value (**"Washington"**) in the list, the variable index will update each time. This means that once the for loop finishes, index will contain the last index location the loop finds (not the first one).


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Create a new list, **five_hundred_list**, that contains only the elements from **crime_rates** that are greater than **500**. To accomplish this, you'll need a for loop and an if statement:
  - The for loop specifies which list we want to iterate over and the name of the iterator variable
  - The if statement determines whether the current element is larger than 500.
  - If the current element is larger than 500, use the **append()** method to add it to **five_hundred_list**.

In [13]:
# put your code here
five_hundred_list = []

for x in crime_rates:
  if x > 500:
    five_hundred_list.append(x)
    
print(five_hundred_list)

[749, 828, 503, 1379, 542, 1405, 835, 1288, 647, 974, 1383, 658, 675, 615, 2122, 587, 543, 563, 992, 1185, 617, 734, 1263, 784, 575, 598, 1750, 1172, 1294, 992, 522, 1216, 815, 639, 1154, 1993, 919, 594, 1160, 636, 752, 517, 738, 503, 704, 597, 1776, 722, 1548, 616, 1171, 724, 990, 1177, 742]


Now that we know how to combine if statements and for loops, we can find the highest crime rate in the **crime_rates** list.

Here's one way we can approach this task:

- Assign the value at index **0** in **crime_rates** to a new integer variable called **highest**
- Use a for loop that compares each value in **crime_rates** to **highest**, and assigns that value to **highest** if it's larger
- If the first value (at index **0**) is the largest value in the list, then the value assigned to **highest** will never change. Otherwise, each time the loop finds a value that's larger than **highest**, it will assign that value to **highest**. The value assigned to **highest** only changes when a new value in the list is larger than it. When the loop completes, we're guaranteed to have the largest value.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Find the largest integer in **crime_rates** using the strategy we just outlined, and assign that value to the variable **highest**.

In [3]:
# put your code here
print(crime_rates)

highest = crime_rates[0]

for number in crime_rates:
  if number > highest:
    highest = number
    
print(highest)
  


[749, 371, 828, 503, 1379, 425, 408, 542, 1405, 835, 1288, 647, 974, 1383, 455, 658, 675, 615, 2122, 423, 362, 587, 543, 563, 168, 992, 1185, 617, 734, 1263, 784, 352, 397, 575, 481, 598, 1750, 399, 1172, 1294, 992, 522, 1216, 815, 639, 1154, 1993, 919, 594, 1160, 636, 752, 130, 517, 423, 443, 738, 503, 413, 704, 363, 401, 597, 1776, 722, 1548, 616, 1171, 724, 990, 169, 1177, 742]
2122


# 4.0 Challenge: Files, Loops, and Conditional Logic

Our notebooks are structured around learning by doing, so you're ready to tackle problems in the real world. We've also designed challenges that will give you an opportunity to practice programming and data science by completing structured problems. Completing challenges will help you solidify concepts and apply what you've learned in the **real world**. 

**We encourage you to think, make your own attempts, and experiment for a significant amount of time before doing so!**

## 4.1 Unisex name

For this challenge, you'll be working with the data set behind this [FiveThirtyEight](http://fivethirtyeight.com/features/there-are-922-unisex-names-in-america-is-yours-one-of-them/) article on common unisex (gender-neutral) names in the United States. You'll start by reading in the file and iteratively converting the data to more useful representations. At the end of this challenge, you'll filter the data so that it only includes the names that at least 1,000 people share.


The staff at [FiveThirtyEight](http://fivethirtyeight.com/) compiled this data set from information at the [Social Security Adminstration's website](https://www.ssa.gov/oact/babynames/limits.html). You'll work with a shortened version of the [full data](https://github.com/fivethirtyeight/data/blob/master/unisex-names/unisex_names_table.csv) set to complete this challenge.

Here's a preview of the shortened data set, which is in a CSV file named **dq_unisex_names.csv**:

>```python
Casey,176544.328149
Riley,154860.66517300002
Jessie,136381.830656
Jackie,132928.78874000002
Avery,121797.41951600001
Jaime,109870.18729000002
Peyton,94896.39521599999
```

Each line contains two values separated by commas. The first value is the **unisex name**, and the second value is the **estimated number of Americans** with that name.

## 4.2 Read the File Into a String

To work with the data, you'll first need to read it into Python. This involves creating a File object, then using one of its methods to read the file into a string.


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Use the **open()** function with the following parameters to return a File object:

  - **dq_unisex_names.csv** for the file name
  - **r** for read mode

2. Then, use the **read()** method of the File object to read the file into a string.

3. Assign that string to a variable named **names**.

In [38]:
# put your code here
f = open("dq_unisex_names.csv", "r")

names = f.read()
print(names)

Casey,176544.328149
Riley,154860.66517300002
Jessie,136381.830656
Jackie,132928.78874000002
Avery,121797.41951600001
Jaime,109870.18729000002
Peyton,94896.39521599999
Kerry,88963.92625
Jody,80400.519199
Kendall,79210.873961
Payton,64151.630388
Skyler,53486.390419
Frankie,51288.068109
Pat,44781.602373
Quinn,41920.940058
Harley,41237.565743
Reese,36360.520613
Robbie,32636.047648000003
Tommie,29528.793818000002
Justice,27350.56457
Kris,24955.811341999997
Carey,24790.097053999998
Emerson,24167.154361
Blair,23160.101097
Amari,22296.161372
Elisha,22184.559374
Sage,21336.062278
Emery,18844.118877
Stevie,18665.219122
Rowan,18367.420464
Ollie,17889.750339
Shea,16768.891880000003
Jaylin,15093.192038999998
Phoenix,14841.367875999998
Charley,12546.265781999999
Armani,12541.667755
Devyn,12515.023073999999
Ivory,11908.580714
Kendal,11841.403479999999
Baby,11132.855753
Mckinley,10972.533414
Finley,10194.618867000001
Milan,9971.955937
Tory,9837.844335
Shay,9444.894035
Shiloh,8999.08523
Lavern,8776.346

## 4.3 Convert the String to a List

Now that you have a string representation of the file, convert it to a list of strings.

<img width="500" src="https://drive.google.com/uc?export=view&id=1huJu8SgydBBR_yGzt1QjMAmh8hBF-b-6">

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Use the **split()** method that strings have to split on the new-line delimiter (**"\n"**), and assign the resulting list to **names_list**.
2. Select the first five elements in **names_list**, and assign them to **first_five**.
3. Display **first_five** using the **print()** function.

In [42]:
# put your code here
names_list = names.split("\n")
first_five = names_list[:5]
print(first_five)

['Casey,176544.328149', 'Riley,154860.66517300002', 'Jessie,136381.830656', 'Jackie,132928.78874000002', 'Avery,121797.41951600001']


## 4.4 Convert the List of Strings to a List of Lists

When you displayed the first five elements, you may have noticed that they contained commas. Let's split each string element in **names_list** on the comma, and add the resulting lists to a new list named **nested_list**.


<img width="500" src="https://drive.google.com/uc?export=view&id=1R1t4-zU6Kjr9NCrZ7dV6cZZA0hwvHEl3">


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. Split each element in **names_list** on the comma delimiter (** , **) and append the resulting list to **nested_list**. To accomplish this:

  - Create an empty list and assign it to **nested_list**.
  - Write a for loop that iterates over **names_list**.
  - Within the loop body, run the **split()** method on each element to return a list (assign that list to **comma_list**).
  - Within the loop body, run the **append()** method to add each list (**comma_list**) to **nested_list**.
2. Use the **print()** function to display the first five elements in **nested_list**.

In [45]:
f = open('dq_unisex_names.csv', 'r')
names = f.read()
names_list = names.split('\n')

# put your code here
nested_list = []

for name in names_list:
  comma_list = name.split(",")
  nested_list.append(comma_list)
  
print(nested_list[:5])

[['Casey', '176544.328149'], ['Riley', '154860.66517300002'], ['Jessie', '136381.830656'], ['Jackie', '132928.78874000002'], ['Avery', '121797.41951600001']]


## 4.5 Convert Numerical Values

You now have a list of lists assigned to **nested_list**, where each inner list contains string elements. The second element (the estimated number of people with that name) in each list is a decimal value that you should convert to a float. By converting these values to floats, you'll be able to perform computations on them and analyze the data.


<img width="600" src="https://drive.google.com/uc?export=view&id=1yOaFpW1mtSI4_UmTq_XN0rPVCQLNtJMS">


**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. Create a new list of lists called **numerical_list** where:

  - The element at index **0** for each list is the unisex name (as a string)
  - The element at index **1** for each list is the number of people who share that name (as a float)

  To accomplish this:

  - Create an empty list and assign it to **numerical_list**.
  - Write a for loop that iterates over **nested_list**. In the loop body:
    - Retrieve the element at index **0** and assign it to a variable.
    - Retrieve the element at index **1**, convert it to a float, and assign it to a variable.
    - Create a new list containing these two elements (in the same order).
    - Use the **append()** method to add this new list to **numerical_list**.
    
2. Finally, display the first five elements in **numerical_list**.




In [48]:
# put your code here
numerical_list = []

for x in nested_list:
  name = x[0]
  number = float(x[1])
  new_list = [name, number]
  numerical_list.append(new_list)
  
print(numerical_list[:5])

[['Casey', 176544.328149], ['Riley', 154860.66517300002], ['Jessie', 136381.830656], ['Jackie', 132928.78874000002], ['Avery', 121797.41951600001]]


## 4.6 Filter the List

The data set contains first names shared by at least 100 people. Let's limit it those shared by at least 1,000 people.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. Create a new list of strings called **thousand_or_greater** that only contains the names shared by 1,000 people or more.

  To accomplish this:
    - Create an empty list and assign it to **thousand_or_greater**.
    - Write a for loop that iterates over **numerical_list**.
    - In the loop body, use an if statement to determine if the value at index 1 for that element (which is a list) is greater than or equal to 1000.
    - If the value is greater than or equal to 1000, use the **append()** method to add its name to **thousand_or_greater**.
2. Finally, display the first 10 elements in **thousand_or_greater**.

In [49]:
# put your code here
thousand_or_greater = []

for x in numerical_list:
  if x[1] >= 1000:
    thousand_or_greater.append(x[0])
    
print(thousand_or_greater)

['Casey', 'Riley', 'Jessie', 'Jackie', 'Avery', 'Jaime', 'Peyton', 'Kerry', 'Jody', 'Kendall', 'Payton', 'Skyler', 'Frankie', 'Pat', 'Quinn', 'Harley', 'Reese', 'Robbie', 'Tommie', 'Justice', 'Kris', 'Carey', 'Emerson', 'Blair', 'Amari', 'Elisha', 'Sage', 'Emery', 'Stevie', 'Rowan', 'Ollie', 'Shea', 'Jaylin', 'Phoenix', 'Charley', 'Armani', 'Devyn', 'Ivory', 'Kendal', 'Baby', 'Mckinley', 'Finley', 'Milan', 'Tory', 'Shay', 'Shiloh', 'Lavern', 'Sky', 'Reilly', 'Leighton', 'Santana', 'Arden', 'Campbell', 'Channing', 'Kamari', 'Alva', 'Clair', 'Lavon', 'Kenyatta', 'Britt', 'Vernell', 'Michal', 'Austyn', 'Rian', 'Dominque', 'Remy', 'Jammie', 'Ricki', 'Jordin', 'Trinidad', 'Ryley', 'Carrol', 'Landry', 'Lorin', 'Yael', 'Codi', 'Sloan', 'Jael', 'Kodi', 'Azariah', 'Dakotah', 'Lorenza', 'Artie', 'Maxie', 'Torey', 'Tai', 'Kalani', 'Codie', 'Storm', 'Jaedyn', 'Merritt', 'Allyn', 'Jourdan', 'Yuri', 'Ellery', 'Oakley', 'Gentry', 'Ellison', 'Arin', 'Sol', 'Carrington', 'Dell', 'Carlin', 'Lennie', 'Dr

# 5.0 List Operations


## 5.1 Introduction to the dataset

In this section, we'll look at daily weather data for Los Angeles (L.A.) during 2014. Here's a look at the beginning of **la_weather.csv**, the data set we'll be working:

>```python
Day,Type of Weather
1,Sunny
2,Sunny
3,Sunny
4,Sunny
5,Sunny
6,Rain
7,Sunny
8,Sunny
9,Fog
10,Rain
```

In the previous sections, we split a CSV file into rows. We'll be doing the same thing here, so it's useful to think of each line in the data file as a separate row. Each row contains multiple values separated by a comma (** , ** ).

The first row in our data is the header row, which contains labels for the values beneath them. As the header row indicates, each row has two values:

- **Day** - A number from 1 to 365 indicating the day of the year. - - January 1st is 1, and December 31st is 365.
- **Type of Weather** - The type of weather experienced on that day. The values that may appear here include **Rain, Sunny, Fog, Fog-Rain, or Thunderstorm**.

Our ultimate goal is to count how many times each type of weather occurred over the course of the year. During this mission, we'll learn how to manipulate the data with lists, and make good progress towards that goal. In the next sections, we'll fit all the pieces together and tally up the data.

## 5.2 Parsing the file

Because our data is in a CSV file, we'll need to read the file in before we can work with it. In the earlier sections, we read a CSV file into a list, and we'll do the same here.

To read the file in, we'll need to:

- Open it with the **open()** function. This will return a File object.
- Read the open file into a variable using the **read()** method. This will return in a string.
- Split the data into rows on the newline character (** \n **). This will result in a list.
- Loop through each row, and split each row into a list on the comma character (** , ** ). This will result in a list of lists.

Here's the code we used to parse the **crime_rates.csv** file in an earlier mission:

>```python
f = open("crime_rates.csv", 'r')
data = f.read()
rows = data.split('\n')
full_data = []
for row in rows:
    split_row = row.split(",")
    full_data.append(split_row)
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. Create an empty list and assign to **weather_data**.
2. Open and read in **la_weather.csv**.
3. Split the data on the newline character to convert it to a list of rows.
4. Split each row on the comma and append each list to **weather_data**.





In [7]:
# put your code here
weather_data = []
f = open("la_weather.csv", "r").read()
list_split = f.split("\n")

for x in list_split:
  weather_data.append(x.split(","))
  
print(weather_data)

[['Day', 'Type of Weather'], ['1', 'Sunny'], ['2', 'Sunny'], ['3', 'Sunny'], ['4', 'Sunny'], ['5', 'Sunny'], ['6', 'Rain'], ['7', 'Sunny'], ['8', 'Sunny'], ['9', 'Fog'], ['10', 'Rain'], ['11', 'Sunny'], ['12', 'Sunny'], ['13', 'Sunny'], ['14', 'Sunny'], ['15', 'Sunny'], ['16', 'Fog'], ['17', 'Sunny'], ['18', 'Sunny'], ['19', 'Sunny'], ['20', 'Sunny'], ['21', 'Sunny'], ['22', 'Sunny'], ['23', 'Rain'], ['24', 'Fog-Rain'], ['25', 'Rain'], ['26', 'Fog-Rain'], ['27', 'Rain'], ['28', 'Sunny'], ['29', 'Sunny'], ['30', 'Sunny'], ['31', 'Sunny'], ['32', 'Sunny'], ['33', 'Rain'], ['34', 'Sunny'], ['35', 'Fog'], ['36', 'Fog'], ['37', 'Sunny'], ['38', 'Sunny'], ['39', 'Rain'], ['40', 'Sunny'], ['41', 'Sunny'], ['42', 'Sunny'], ['43', 'Sunny'], ['44', 'Sunny'], ['45', 'Fog'], ['46', 'Sunny'], ['47', 'Sunny'], ['48', 'Fog'], ['49', 'Sunny'], ['50', 'Rain'], ['51', 'Sunny'], ['52', 'Sunny'], ['53', 'Sunny'], ['54', 'Sunny'], ['55', 'Sunny'], ['56', 'Sunny'], ['57', 'Sunny'], ['58', 'Sunny'], ['59', '

## 5.3 Getting a Single Column From the Data

We've been thinking of our data in terms of rows, where rows are horizontal, from left to right:

<img width="200" src="https://drive.google.com/uc?export=view&id=15yFHKRos6slDWcjU8VpUTawT2YV3vz66">

The diagram above shows the first four rows in the data, along with their corresponding row numbers. Each item in a row records different information, but refers to the same day.

Another way to think of the data is in terms of values that are alike, or columns. Columns are vertical, from top to bottom:

<img width="250" src="https://drive.google.com/uc?export=view&id=1L-i0qbtRk7rEdgPHtL1TLQNTwqmJ4hsl">


In the above diagram, we grouped the data by column. Each value in a column records similar information. Column **1** shows the **Day of the year**, and column **2** shows the **Type of Weather** that occurred on that day.

The first value in each column is from the header row; it labels the data that appears in the column. While not all data sets include headers, they're helpful to have.

Since we'll be counting the total number of times each type of weather occurred during the year, we don't need the **Day** column.

You may recall that in a previous mission, we did the following to extract a column:

- Looped over each row.
- Retrieved the second element of the row.
- Appended the second element to a list of items.

Here's an example of what this process looks like:

>```python
numbers = [[1,2],[3,4],[5,6],[7,8]]
second_column = []
for item in numbers:
    value = item[1]
    second_column.append(value)
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Loop over each row in **weather_data**.
2. Append the second item in each row to the **weather** list.
3. When complete, **weather** should contain each value from the **Type of Weather** column.

In [8]:
# put your code here
weather = []
for valor in weather_data:
  weather.append(weather_data[1])
print(weather)

[['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', 'Sunny'], ['1', '

## 5.4 Counting the Items in a List

To verify that we selected the **Type of Weather** column properly, let's tally the number of items in **weather**.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Count the number of items in **weather**. You can accomplish this by:

  - Looping over each element in **weather.**
  - Adding **1** to **count** for each element.
2. When finished, **count** should equal the number of items in **weather.**

In [9]:
# put your code here
cont = 0
for x in weather:
  cont += 1
print(cont)

366


## 5.6 Removing the Header

When we created the **weather** list, we included the header. The first few items in **weather** look like this:

<img width="600" src="https://drive.google.com/uc?export=view&id=1hGsmarMYgcF8Bws4ckuhc2jriai5ehsD">


We want to tally how many times each type of weather occurred in L.A. To do this, we'll need to remove the string **Type of Weather** from the beginning of our list. If we don't do this, our count will include the header, which isn't part of the data set. The header is just a string that tells us what kind of values are in a column.

To remove the header we can use a list slicing. **weather** contains 366 elements, and the last list index value is 365. **We'll want to create a slice that goes from the second element (index 1) to the end of the list.**

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Slice the **weather** list to remove the header.
2. The slice should only remove the first element in the list.
3. Assign the slice to **new_weather.**

In [10]:
new_weather = weather.pop(0)
print(new_weather)


['1', 'Sunny']


## 5.7 The In Statement

In Python, it's often very useful to check whether a certain element is present in a list. One way to perform this check is to use a for loop:

>```python
animals = ["cat", "dog", "rabbit"]
for animal in animals:
    if animal == "cat":
        print("Cat found")
```

This is a lot of code for a common operation, though. Luckily, Python has the in statement, which allows us to check whether a list contains a specific element much more quickly:

>```python
animals = ["cat", "dog", "rabbit"]
if "cat" in animals:
    print("Cat found")
```

The **in statement** checks whether certain value occurs within a list, and returns **True** if it does. If not, the statement returns **False**.

We can directly assign the result to a variable as well:

>```python
animals = ["cat", "dog", "rabbit"]
cat_found = "cat" in animals
```

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">


1. Use the **in statement** to check whether the value **cat** is in the list **animals**, and assign the result to **cat_found**.
2. Use the **in statement** to check whether the value **"space_monster"** is in the list **animals**, and assign the result to **space_monster_found**.

In [66]:
animals = ["cat", "dog", "rabbit", "horse", "giant_horrible_monster"]

# put your code here
cat_found = "cat" in animals
space_monster_found = "space_monster" in animals

print(cat_found, space_monster_found)

True False


## 5.8 Weather Types

We can use the **in statement** to discover which types of weather the **new_weather** list contains. At the start of this section, we specified that the data set includes **Rain, Sunny, Fog, Fog-Rain, and Thunderstorm**. Let's verify that each of these types occurs in **new_weather**.

**Exercise**

<img width="60" src="https://drive.google.com/uc?export=view&id=1QoTRiOtUzjnbRL7Ue5uPxKse03tE1tPe">

1. Loop through each item in the **weather_types** list.
2. Check whether the item occurs in the **new_weather** list.
3. Append the result of the check to **weather_type_found**.
4. At the end, **weather_type_found** should be a list of Boolean values.

In [11]:
weather_types = ["Rain", "Sunny", "Fog", "Fog-Rain", "Thunderstorm", "Type of Weather"]
weather_type_found = []

# put your code here
for item in weather_types:
  if item in new_weather:
    weather_type_found.append(True)
  else:
    weather_type_found.append(False)
    
print(weather_type_found)

[False, True, False, False, False, False]
